warehouse/migrations/versions/5ea52744d154_add_macaroons.py

```diff
@@ -0,0 +1,58 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Add Macaroons
+
+Revision ID: 5ea52744d154
+Revises: f2a453c96ded
+Create Date: 2018-07-16 06:45:31.152291
+"""
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+revision = "5ea52744d154"
+down_revision = "f2a453c96ded"
+
+
+def upgrade():
+    op.create_table(
+        "macaroons",
+        sa.Column(
+            "id",
+            postgresql.UUID(as_uuid=True),
+            server_default=sa.text("gen_random_uuid()"),
+            nullable=False,
+        ),
+        sa.Column("user_id", postgresql.UUID(as_uuid=True), nullable=False),
+        sa.Column(
+            "description", sa.String(length=100), server_default="", nullable=False
+        ),
+        sa.Column(
+            "created", sa.DateTime(), server_default=sa.text("now()"), nullable=False
+        ),
+        sa.Column("last_used", sa.DateTime(), nullable=True),
+        sa.Column(
+            "caveats",
```

Comment:
What is the rationale of storing root caveats explicitly?
IIRC:
* root macaroon == f(key, caveats)
* key must be stored (somewhere) to allow validation
* id is needed to find the key and to invalidate macaroon tree

Thus, (id, key, root macaroon (bytes) ) is all the information needed, right?
Is the idea to recompute root macaroon on demand?
Can root caveats be edited?
What's validates, caveats in request macaroon or in this table?

Reply:
The only reason to store the caveats in the database is to make it possible for the UI to show what the maximum scope of this token is. Of course given the way macaroons function we can't tell them what derived macaroons they've made from that initial token and what those are scoped to, but we can tell them what the *max* scope their token is good for.

We would never use what's in the database here for the validation, it'd just be there for informational purposes.